### PR TITLE
Added hashing trick nodes

### PR DESCRIPTION
**Pull request by tomerk**
issue #184

### CODE REVIEW

File: src/main/scala/nodes/nlp/NGramsHashingTF.scala [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+/**
+ * Converts the n-grams of a sequence of terms to a sparse vector representing their frequencies,
+ * using the hashing trick: https://en.wikipedia.org/wiki/Feature_hashing
+ *
+ * It computes a rolling MurmurHash3 instead of fully constructing the n-grams, making
+ * it more efficient than using [[NGramsFeaturizer]] followed by [[HashingTF]], although
+ * it should return the exact same feature vector.
+ *
+ * Individual terms are hashed using Scala's `.##` method. We may want to convert to MurmurHash3 for strings,
```
Comment (etrain) [inline src/main/scala/nodes/nlp/NGramsHashingTF.scala]: Is this comment still relevant?

File: src/main/scala/nodes/nlp/NGramsHashingTF.scala [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    case _ =>
+  }
+
+  final val seqSeed = "Seq".hashCode
+
+  /** Mix in a block of data into an intermediate hash value. */
+  final def mix(hash: Int, data: Int): Int = {
+    var h = mixLast(hash, data)
+    h = rotl(h, 13)
+    h * 5 + 0xe6546b64
```
Comment (etrain) [inline src/main/scala/nodes/nlp/NGramsHashingTF.scala]: Can we get some comments about where these constants come from?

File: src/test/scala/nodes/nlp/NGramsHashingTFSuite.scala [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  test("NGramsHashingTF 1 to 1") {
+    val dims = 40000
+
+    val testDatum = "this sentence is a sentence is the some there some then there some".split(" ")
+    val ngrams = NGramsFeaturizer(1 to 1).apply(testDatum)
+    val tfVector = HashingTF(dims).apply(ngrams)
+
+    val ngramsHashingTFVector = NGramsHashingTF(1 to 1, dims).apply(testDatum)
+
+    // Assert that the NGramsHashingTF node returns the same output as first getting n-grams then hashing
```
Comment (etrain) [inline src/test/scala/nodes/nlp/NGramsHashingTFSuite.scala]: yes, i like this test!

**Comment by etrain** [comment]
This looks good to me - if you could clean up the comments and add a direct reference to the Murmur Hash 3 algorithm. Is there a reason we're reimplementing rather than linking to the one in the scala stdlib? 

http://www.scala-lang.org/api/2.10.4/#scala.util.hashing.MurmurHash3$

**Comment by tomerk** [comment]
Yeah the code (and constants) come directly from the scala stdlib, but the scala stdlib marked these rolling hash methods as `private[hashing]`. I could make a util in the hashing namespace that exposes these methods to us, which may be preferable to the code & constants copying I've done here.

File: src/main/scala/nodes/nlp/NGramsHashingTF.scala (hunk 1 above)
Comment (tomerk) [inline src/main/scala/nodes/nlp/NGramsHashingTF.scala]: Yeah, the terms' hash values are combined together using MurmurHash3 to form the n-gram hashes, but the individual terms are still hashed using the `.##` method.

**Comment by etrain** [comment]
Copy/pasting is fine, then, I guess. Let's just leave a reference to where it came from in the code?

**Comment by tomerk** [comment]
Okay @etrain I added a reference to the scala stdlib class.

**Comment by etrain** [comment]
LGTM, merging this!

**Pull request merged**